src/main/java/baritone/cache/FasterWorldScanner.java

```diff
@@ -0,0 +1,289 @@
+import baritone.utils.accessor.IBitArray;
+import baritone.utils.accessor.IBlockStateContainer;
+import io.netty.buffer.Unpooled;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.BitArray;
+import net.minecraft.util.ObjectIntIdentityMap;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.chunk.*;
+import net.minecraft.world.chunk.storage.ExtendedBlockStorage;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.IntConsumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public enum FasterWorldScanner implements IWorldScanner {
+    INSTANCE;
+    @Override
+    public List<BlockPos> scanChunkRadius(IPlayerContext ctx, BlockOptionalMetaLookup filter, int max, int yLevelThreshold, int maxSearchRadius) {
+        return new WorldScannerContext(filter, ctx).scanAroundPlayerRange(maxSearchRadius);
+    }
+
+    @Override
+    public List<BlockPos> scanChunk(IPlayerContext ctx, BlockOptionalMetaLookup filter, ChunkPos pos, int max, int yLevelThreshold) {
+        return new WorldScannerContext(filter, ctx).scanAroundPlayerUntilCount(max);
+    }
+
+    @Override
+    public int repack(IPlayerContext ctx) {
+        return this.repack(ctx, 40);
+    }
+
+    @Override
+    public int repack(IPlayerContext ctx, int range) {
+        IChunkProvider chunkProvider = ctx.world().getChunkProvider();
+        ICachedWorld cachedWorld = ctx.worldData().getCachedWorld();
+
+        BetterBlockPos playerPos = ctx.playerFeet();
+
+        int playerChunkX = playerPos.getX() >> 4;
+        int playerChunkZ = playerPos.getZ() >> 4;
+
+        int minX = playerChunkX - range;
+        int minZ = playerChunkZ - range;
+        int maxX = playerChunkX + range;
+        int maxZ = playerChunkZ + range;
+
+        int queued = 0;
+        for (int x = minX; x <= maxX; x++) {
+            for (int z = minZ; z <= maxZ; z++) {
+                Chunk chunk = chunkProvider.getLoadedChunk(x, z);
+
+                if (chunk != null && !chunk.isEmpty()) {
+                    queued++;
+                    cachedWorld.queueForPacking(chunk);
+                }
+            }
+        }
+
+        return queued;
+    }
+
+    // for porting, see {@link https://github.com/JsMacros/JsMacros/blob/backport-1.12.2/common/src/main/java/xyz/wagyourtail/jsmacros/client/api/classes/worldscanner/WorldScanner.java}
+    // tho I did change some things...
+    public static class WorldScannerContext {
+        private final BlockOptionalMetaLookup filter;
+        private final IPlayerContext ctx;
+        private final Map<IBlockState, Boolean> cachedFilter = new ConcurrentHashMap<>();
```

Comment:
Appendix to previous comment since I seem to not have saved my edit before submitting:
A `HashSet<Block>` is probably small enough to precompute it for every `BlockOptinalMetaLookup`, which is the reason why I suggested using it rather than a `HashSet<IBlockState>`. The latter might still provide an advantage if used for scanning, possibly multiple times, and is maybe maybe still small enough for eager population in those cases (it's up to 16 times bigger on 1.12.2 and even more on newer versions).

Reply:
to my eyes, a ConcurrentHashMap where the value is a Boolean instantly sets off alarm bells lmao

will write another comment after actually reading what it's for :P

Reply:
I think a better approach would follow something like https://github.com/cabaletta/baritone/pull/3479/files#diff-4cd83dc955fbd7b9fd34a9d639818d599e947605ae50cedb5dcf939a1b0daec6R59

Benefits include 1. there's no mutex so there's no locking or thread contention slowdown 2. there's no new hashmap that blows out a ram cache line for each entry, instead a compact int array that fits in cache easily

Reply:
so with that solution, there would be an extra lookup operation every time to convert the blockstate to a blockstate id...

Reply:
also, iirc, the concurrent part means it at least uses a read-write lock